eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToList(inputArray) { 
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let currentIndex = inputArray.length-1;
+
+    function innerArrayToList(inputArray, list = null) {
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        let arrayElement = inputArray[currentIndex];
+        currentIndex--;
+        return innerArrayToList(inputArray, { value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(inputArray);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        let element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest:listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    let arrayAccumulator = [];
+
+    function innerListToArray(list) {
+        arrayAccumulator.push(list.value);
+        if (list.rest) {
+            innerListToArray(list.rest)
+        }
+        return arrayAccumulator;
+    }
+
+    return innerListToArray(inputList);
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArrayIterative(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+
+    let listLength = getListLength(inputList);
+    
+    let acc = [];
+    let listBuffer = inputList;
+    for (let i = 0; i < listLength; i++) {
+        if(listBuffer.value !== null) {
+            acc.push(listBuffer.value);
+        }
+        listBuffer = listBuffer.rest;
+    }
+
+    return acc;
+}
+
+/**
+ * Returns list from input list with prepended element
+ * @param {*} inputList 
+ * @param {*} element 
+ * @returns 
+ */
+function prepend(inputList, element) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    return {
+        value: element,
+        rest: inputList
+    }
+}
+
+/**
+ * Returns nth element in list
+ * @param {*} inputList 
+ * @param {*} targetIndex 
+ * @returns 
+ */
+function nth(inputList, targetIndex) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    if (isNaN(targetIndex) || targetIndex < 0 || Math.abs(targetIndex) === Infinity) {
+        throw "invalid target index";
+    }
+
+    listBuffer = inputList;
+    for (let i = 0; i < targetIndex; i++) {
```

Comment:
what if targetIndex is greater than the length of input list? It's fine if you don't handle it, but if that's the case you should put it in the docstring so callers will be aware